core/finality-grandpa/src/lib.rs

```diff
@@ -216,6 +215,30 @@ impl From<ClientError> for Error {
 	}
 }
 
+/// A stream used by NetworkBridge in its implementation of Network.
+pub struct NetworkStream {
+	inner: Option<mpsc::UnboundedReceiver<Vec<u8>>>,
+	outer: oneshot::Receiver<mpsc::UnboundedReceiver<Vec<u8>>>
+}
+
+impl Stream for NetworkStream {
+	type Item = Vec<u8>;
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+		if let Some(ref mut inner) = self.inner {
+			return inner
+				.poll()
+		}
+		if let Ok(futures::Async::Ready(mut inner)) = self.outer.poll() {
+			let poll_result = inner.poll();
+			self.inner = Some(inner);
+			return poll_result
+		}
+		Ok(futures::Async::NotReady)
```

Comment:
Maybe rewrite this whole function as:
```
loop {
    match self.inner {
        Some(ref mut inner) => return inner.poll(),
        None => match self.outer.poll() {
            Ok(Ready(inner)) => self.inner = Some(inner),
            Ok(NotReady) => return Ok(NotReady),
            Err(e) => handle the error,
       }
    }
}
```

Reply:
clever, and I'd rather keep it simple and rely on the next call to `poll`.